packages/next/src/server/route-handlers/app-route/app-route-route-handler.ts

```diff
@@ -0,0 +1,320 @@
+    duplex: 'half',
+    method: req.method,
+    headers: fromNodeHeaders(req.headers),
+  })
+}
+
+function resolveHandlerError(err: any): Response {
+  if (isRedirectError(err)) {
+    const redirect = getURLFromRedirectError(err)
+    if (!redirect) {
+      throw new Error('Invariant: Unexpected redirect url format')
+    }
+
+    // This is a redirect error! Send the redirect response.
+    return handleTemporaryRedirectResponse(redirect)
+  }
+
+  if (isNotFoundError(err)) {
+    // This is a not found error! Send the not found response.
+    return handleNotFoundResponse()
+  }
+
+  // TODO: validate the correct handling behavior
+  Log.error(err)
+  return handleInternalServerErrorResponse()
+}
+
+async function sendResponse(
+  req: BaseNextRequest,
+  res: BaseNextResponse,
+  response: Response
+): Promise<void> {
+  // Copy over the response status.
+  res.statusCode = response.status
+  res.statusMessage = response.statusText
+
+  // Copy over the response headers.
+  response.headers.forEach((value, name) => {
+    // The append handling is special cased for `set-cookie`.
+    if (name.toLowerCase() === 'set-cookie') {
+      res.setHeader(name, value)
+    } else {
+      res.appendHeader(name, value)
+    }
+  })
+
+  /**
+   * The response can't be directly piped to the underlying response. The
+   * following is duplicated from the edge runtime handler.
+   *
+   * See packages/next/server/next-server.ts
+   */
+
+  const originalResponse = (res as NodeNextResponse).originalResponse
+
+  // A response body must not be sent for HEAD requests. See https://httpwg.org/specs/rfc9110.html#HEAD
+  if (response.body && req.method !== 'HEAD') {
+    const { consumeUint8ArrayReadableStream } =
+      require('next/dist/compiled/edge-runtime') as typeof import('next/dist/compiled/edge-runtime')
+    const iterator = consumeUint8ArrayReadableStream(response.body)
+    try {
+      for await (const chunk of iterator) {
+        originalResponse.write(chunk)
+      }
+    } finally {
+      originalResponse.end()
+    }
+  } else {
+    originalResponse.end()
+  }
+}
+
+/**
+ *
+ */
+export class AppRouteRouteHandler implements RouteHandler<RouteKind.APP_ROUTE> {
+  constructor(
+    private readonly requestAsyncLocalStorageWrapper: AsyncStorageWrapper<
+      RequestStore,
+      RequestContext
+    > = new RequestAsyncStorageWrapper(),
+    private readonly moduleLoader: ModuleLoader = new NodeModuleLoader()
+  ) {}
+
+  private resolve(
+    req: BaseNextRequest,
+    mod: AppRouteModule
+  ): AppRouteHandlerFn {
+    // Ensure that the requested method is a valid method (to prevent RCE's).
+    if (!isHTTPMethod(req.method)) return handleBadRequestResponse
+
+    // Pull out the handlers from the app route module.
+    const { handlers } = mod
+
+    // Check to see if the requested method is available.
+    const handler: AppRouteHandlerFn | undefined = handlers[req.method]
```

Comment:
Did you handle the lowercase method? Only uppercase is supported but it should throw a helpful error when someone uses the lowercase version.

Reply:
I can add support for that, but it would only happen during development, is that ok?